spring-integration-core/src/main/java/org/springframework/integration/handler/advice/RequestHandlerRateLimiterAdvice.java

```diff
@@ -0,0 +1,114 @@
+import org.springframework.util.Assert;
+
+/**
+ * Rate limiter advice, which will ensure call to service is Rate limited within set rate limit.
+ * Request calls breaching the rate limit will wait for time rate limit is not breached.
+ *
+ * @author Meherzad Lahewala
+ * @since 5.1
+ */
+public class RequestHandlerRateLimiterAdvice extends AbstractRequestHandlerAdvice {
+
+	private final AtomicInteger currentIndex;
+	private final long maxDelayInMillis;
+	private final long oneTimeUnitInMillis;
+	private final AtomicLongArray queue;
+	private final int rate;
+
+	/**
+	 * Create an instance of RequestHandlerRateLimiterAdvice with the provided arguments.
+	 *
+	 * @param rate the rate per time unit.
+	 * @param maxDelay the max allowable delay in time units. Use -1 for no max delay.
+	 * @param timeUnit the time unit.
+	 */
+	public RequestHandlerRateLimiterAdvice(int rate, int maxDelay, TimeUnit timeUnit) {
+		Assert.isTrue(rate > 0, "Rate should be a positive integer");
+		Assert.notNull(timeUnit, "Timeunit should be non null");
+		this.currentIndex = new AtomicInteger(0);
+		this.maxDelayInMillis = timeUnit.toMillis(maxDelay);
+		this.oneTimeUnitInMillis = timeUnit.toMillis(1);
+		this.queue = new AtomicLongArray(rate);
+		this.rate = rate;
+	}
+
+	/**
+	 * Create an instance of RequestHandlerRateLimiterAdvice with the provided arguments.
+	 *
+	 * @param rate the rate per time unit.
+	 * @param timeUnit the time unit.
+	 */
+	public RequestHandlerRateLimiterAdvice(int rate, TimeUnit timeUnit) {
+		this(rate, -1, timeUnit);
+	}
+
+	@Override
+	protected Object doInvoke(ExecutionCallback callback, Object target, Message<?> message) throws Exception {
+		delayIfNecessary(target, message);
+		return callback.execute();
+	}
+
+	private void delayIfNecessary(Object target, Message<?> message) throws InterruptedException {
+		long timeToWait = -1;
+		while (timeToWait == -1) {
+			long now = System.currentTimeMillis();
+			int index = this.currentIndex.getAndUpdate(value -> value + 1 < this.rate ? value + 1 : 0);
+			long oldValue = this.queue.get(index);
+			long waitTill;
+			if (now < oldValue) {
+				waitTill = oldValue + this.oneTimeUnitInMillis;
+			}
+			else {
+				long diff = now - oldValue;
+				waitTill = now + (diff < this.oneTimeUnitInMillis ? this.oneTimeUnitInMillis - diff : 0);
+			}
+			if (this.queue.compareAndSet(index, oldValue, waitTill)) {
+				timeToWait = waitTill - now;
+			}
+		}
+		if (this.maxDelayInMillis != -1 && timeToWait > this.maxDelayInMillis) {
+			throw new RateLimiterMaxDelayException(message, "Rate limiter maxDelay exceeded for " + target);
+		}
+		if (logger.isDebugEnabled() && timeToWait > 0) {
+			logger.debug(String.format("Sleeping for %s ms", timeToWait));
+		}
+		TimeUnit.MILLISECONDS.sleep(timeToWait);
+	}
+
+
+	/**
+	 * An exception thrown when the max delay is triggered for rate limiter.
+	 */
+	public static final class RateLimiterMaxDelayException extends MessagingException {
```

Comment:
Well, actually exceptions are named after actions: `ResourceNotFound`, `TransactionTimeout` etc.

I'm not sure in the algorithm here yet, so I can't come up with the better name...

`MaxDelayExceededException` ?

The `RateLimiter` is already implied with the wrapping class. 🤷‍♂️